samples/monte_carlo.py

```diff
@@ -0,0 +1,519 @@
+
+for _ in range(N_acid):
+    pos = np.random.random(3) * system.box_l
+    system.part.add(pos=pos, type=types["H+"], q=charges["H+"])
+
+for type_pair in itertools.combinations_with_replacement(types.values(), 2):
+    system.non_bonded_inter[type_pair[0], type_pair[1]].wca.set_params(
+        epsilon=epsilon, sigma=sigma)
+
+p3m = espressomd.electrostatics.P3M(
+    prefactor=2., accuracy=1e-2, mesh=8, cao=3, verbose=False)
+dh = espressomd.electrostatics.DH(
+    prefactor=2., kappa=0., r_cut=0.2 * box_l)
+
+# energy minimize the system
+system.integrator.set_steepest_descent(
+    f_max=0., gamma=0.1, max_displacement=0.1 * sigma)
+system.integrator.run(200)
+system.electrostatics.solver = p3m
+system.integrator.run(1000)
+
+# thermalize the system
+system.integrator.set_vv()
+system.thermostat.set_langevin(kT=kT, gamma=friction, seed=SEED)
+system.integrator.run(1000)
+system.electrostatics.solver = dh
+
+if args.mode == "core":
+    if args.method == "cph":
+        ReactionMethod = espressomd.reaction_methods.ConstantpHEnsemble
+    elif args.method == "re":
+        ReactionMethod = espressomd.reaction_methods.ReactionEnsemble
+elif args.mode == "python":
+    if args.method == "cph":
+        ReactionMethod = ConstantpHEnsemble
+    elif args.method == "re":
+        ReactionMethod = ReactionEnsemble
+    params["system"] = system
+
+# set up reaction method
+RE = ReactionMethod(**params)
+RE.set_non_interacting_type(type=max(types.values()) + 1)
+if args.method == "cph":
+    RE.add_reaction(
+        gamma=10**-pKa,
+        reactant_types=[types["HA"]],
+        product_types=[types["A-"], types["H+"]],
+        default_charges={types[name]: charges[name] for name in types.keys()})
+elif args.method == "re":
+    RE.add_reaction(
+        gamma=1e-3,
+        reactant_types=[types["HA"]],
+        reactant_coefficients=[1],
+        product_types=[types["A-"], types["H+"]],
+        product_coefficients=[1, 1],
+        default_charges={types[name]: charges[name] for name in types.keys()})
+reaction = RE.reactions[0]
+system.setup_type_map(type_list=list(types.values()))
+
+# equilibrate the polyelectrolyte chains
+for i in range(5):
+    RE.reaction(steps=10 * N_steps_MC)
+    system.integrator.run(N_steps_MD)
+
+
+@profile
+def sample_alpha(length):
+    alpha_list = []
+    for _ in tqdm.tqdm(range(length)):
+        system.integrator.run(steps=N_steps_MD)
+        RE.reaction(steps=N_steps_MC)
+        alpha = system.number_of_particles(type=types["A-"]) / N_acid
+        alpha_list.append(alpha)
+    return alpha_list
+
+
+sample_size = 100
+tick = time.time()
+alphas = sample_alpha(sample_size)
+tock = time.time()
+
+alpha_avg = np.mean(alphas)
+alpha_err = 1.96 * np.sqrt(np.var(alphas) / len(alphas))
+acceptance_rate = reaction.get_acceptance_rate()
+print(f"acceptance rate = {100. * acceptance_rate:.0f}%")
+print(f"alpha = {alpha_avg:.2f} +/- {alpha_err:.2f}")
+print(f"runtime = {tock - tick:.2f}s")
```

Comment:
What do you think about splitting this into yet another python module to separate the example and the mc functionality? 

Did you already consider using just in time compilation in some of the functions to cut down the factor 4 difference?